# Import the libraries:


In [15]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import pickle
from pandas_profiling import ProfileReport
from sklearn.linear_model import LinearRegression

# Get the Data:

In [6]:
df = pd.read_csv("Advertising.csv")
df.head(5)

Unnamed: 0     TV  Radio  Newspaper  Sales
0           1  230.1   37.8       69.2   22.1
1           2   44.5   39.3       45.1   10.4
2           3   17.2   45.9       69.3    9.3
3           4  151.5   41.3       58.5   18.5
4           5  180.8   10.8       58.4   12.9

# Profiling:

Feature Column:

# Build Model: (###with three features)

In [11]:
data_x = df[['TV','Radio','Newspaper']]

In [13]:
y = df.Sales

In [16]:
lm = LinearRegression()

In [17]:
lm.fit(data_x,y)

LinearRegression()

In [18]:
lm.coef_

array([ 0.04576465,  0.18853002, -0.00103749])

In [19]:
lm.intercept_

2.9388893694594067

y = 0.04TV + 0.18 Radio - 0.001 Newspaper + 2.93

In [22]:
lm.score(data_x,y)

0.8972106381789522

# Build Model with 2 Features:

In [28]:
data_x2 = df[['TV','Radio']]

In [29]:
lm2 = LinearRegression()

In [30]:
lm2.fit(data_x2,y)

LinearRegression()

In [31]:
lm2.score(data_x2,y)

0.8971942610828956

#good score = Feture selection is proper

# Build Model with 2 Features:

In [32]:
data_x3 = df[['TV','Newspaper']]

In [33]:
lm3 = LinearRegression()

In [37]:
lm3.fit(data_x3,y)

LinearRegression()

In [38]:
lm3.score(data_x3,y)

0.6458354938293271

# How to do the feature selection?

In [39]:
lm.coef_

array([ 0.04576465,  0.18853002, -0.00103749])

In [43]:
import statsmodels.formula.api as smf
lm = smf.ols(formula ='Sales ~ TV', data = df).fit()
lm.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Sales   R-squared:                       0.612
Model:                            OLS   Adj. R-squared:                  0.610
Method:                 Least Squares   F-statistic:                     312.1
Date:                Thu, 28 Apr 2022   Prob (F-statistic):           1.47e-42
Time:                        19:41:24   Log-Likelihood:                -519.05
No. Observations:                 200   AIC:                             1042.
Df Residuals:                     198   BIC:                             1049.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      7.0326      0.458     15.360      0.000       6.130       7.935
TV             0.0475      0.003     17.668      0.000       0.042       0.053
==============================================================================
Omnibus:                        0.531   Durbin-Watson:                   1.935
Prob(Omnibus):                  0.767   Jarque-Bera (JB):                0.669
Skew:                          -0.089   Prob(JB):                        0.716
Kurtosis:                       2.779   Cond. No.                         338.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [44]:
import statsmodels.formula.api as smf
lm = smf.ols(formula ='Sales ~ TV + Radio', data = df).fit()
lm.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Sales   R-squared:                       0.897
Model:                            OLS   Adj. R-squared:                  0.896
Method:                 Least Squares   F-statistic:                     859.6
Date:                Thu, 28 Apr 2022   Prob (F-statistic):           4.83e-98
Time:                        19:42:26   Log-Likelihood:                -386.20
No. Observations:                 200   AIC:                             778.4
Df Residuals:                     197   BIC:                             788.3
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      2.9211      0.294      9.919      0.000       2.340       3.502
TV             0.0458      0.001     32.909      0.000       0.043       0.048
Radio          0.1880      0.008     23.382      0.000       0.172       0.204
==============================================================================
Omnibus:                       60.022   Durbin-Watson:                   2.081
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              148.679
Skew:                          -1.323   Prob(JB):                     5.19e-33
Kurtosis:                       6.292   Cond. No.                         425.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [45]:
import statsmodels.formula.api as smf
lm = smf.ols(formula ='Sales ~ TV + Radio + Newspaper', data = df).fit()
lm.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Sales   R-squared:                       0.897
Model:                            OLS   Adj. R-squared:                  0.896
Method:                 Least Squares   F-statistic:                     570.3
Date:                Thu, 28 Apr 2022   Prob (F-statistic):           1.58e-96
Time:                        19:42:44   Log-Likelihood:                -386.18
No. Observations:                 200   AIC:                             780.4
Df Residuals:                     196   BIC:                             793.6
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      2.9389      0.312      9.422      0.000       2.324       3.554
TV             0.0458      0.001     32.809      0.000       0.043       0.049
Radio          0.1885      0.009     21.893      0.000       0.172       0.206
Newspaper     -0.0010      0.006     -0.177      0.860      -0.013       0.011
==============================================================================
Omnibus:                       60.414   Durbin-Watson:                   2.084
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              151.241
Skew:                          -1.327   Prob(JB):                     1.44e-33
Kurtosis:                       6.332   Cond. No.                         454.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

P>|t| for newspaper is 0.86 thus we can reject newspaper as feature

# Variance inflation factor = 1/(1-r2) if >10 then there is high correlation: for multicollinearity

# Multicollinearity is wrt feature vs feature and not feature vs label